python_modules/dagster/dagster/daemon/workspace.py

```diff
@@ -68,15 +71,33 @@ def get_repository_location(self, location_name: str) -> RepositoryLocation:
 
         return location_entry.repository_location
 
-    def cleanup(self) -> None:
+    # change this default param
+    def cleanup(self, cleanup_locations: bool = True) -> None:
         if self._location_entries != None:
-            for location_entry in self._location_entries.values():
-                if location_entry.repository_location:
-                    location_entry.repository_location.cleanup()
+            if cleanup_locations:
+                for location_entry in self._location_entries.values():
+                    if location_entry.repository_location:
+                        location_entry.repository_location.cleanup()
             self._location_entries = None
 
     def __exit__(self, exception_type, exception_value, traceback):
-        self.cleanup()
+        self.cleanup(cleanup_locations=True)
+
+
+# A copy of the main workspace's locations that can be called from a background thread
+# in a daemon without worrying that the main thread will clean up the locations underneath us.
+# Analagous to WorkspaceRequestContext in Dagit.
+#
+# Daemons that call this should be careful to set cleanup_locations=False when calling cleanup
+# on the parent workspace that get_workspace_copy_for_iteration() was called on to create
+# this workspace.
+class DaemonIterationWorkspace(BaseDaemonWorkspace):
```

Comment:
nit: prefer this to be a docstring so that editors can see the description on hover/on typeahead